dev-tools/xtask/src/illumos.rs

```diff
@@ -0,0 +1,144 @@
+use anyhow::{bail, Context, Result};
+use camino::Utf8Path;
+use cargo_metadata::Message;
+use fs_err as fs;
+use serde::Deserialize;
+use std::{
+    collections::{BTreeMap, BTreeSet, HashMap},
+    io::BufReader,
+    process::{Command, Stdio},
+};
+
+use crate::load_workspace;
+
+#[derive(Deserialize, Debug)]
+struct LibraryConfig {
+    binary_allow_list: Option<BTreeSet<String>>,
+}
+
+#[derive(Deserialize, Debug)]
+struct XtaskConfig {
+    libraries: BTreeMap<String, LibraryConfig>,
+}
+
+#[derive(Debug)]
+enum LibraryError {
+    Unexpected(String),
+    NotAllowed(String),
+}
+
+/// Verify that the binary at the provided path complies with the rules laid out
+/// in the xtask.toml config file. Errors are pushed to a hashmap so that we can
+/// display to a user the entire list of issues in one go.
+fn verify_executable(
+    config: &XtaskConfig,
+    path: &Utf8Path,
+    errors: &mut HashMap<String, Vec<LibraryError>>,
+) -> Result<()> {
+    let binary = path.file_name().context("basename of executable")?;
+
+    let command = Command::new("elfedit")
+        .arg("-o")
+        .arg("simple")
+        .arg("-r")
+        .arg("-e")
+        .arg("dyn:tag NEEDED")
+        .arg(&path)
+        .output()
+        .context("exec elfedit")?;
+
+    if !command.status.success() {
+        bail!("Failed to execute elfedit successfully {}", command.status);
+    }
+
+    let stdout = String::from_utf8(command.stdout)?;
+    // `elfedit -o simple -r -e "dyn:tag NEEDED" /file/path` will return
+    // a new line seperated list of required libraries so we walk over
+    // them looking for a match in our configuration file. If we find
+    // the library we make sure the binary is allowed to pull it in via
+    // the whitelist.
+    for library in stdout.lines() {
+        let library_config = match config.libraries.get(library.trim()) {
+            Some(config) => config,
+            None => {
+                errors
+                    .entry(binary.to_string())
+                    .or_default()
+                    .push(LibraryError::Unexpected(library.to_string()));
+
+                continue;
+            }
+        };
+
+        if let Some(allowed) = &library_config.binary_allow_list {
+            if !allowed.contains(binary) {
+                errors
+                    .entry(binary.to_string())
+                    .or_default()
+                    .push(LibraryError::NotAllowed(library.to_string()));
+            }
+        }
+    }
+
+    Ok(())
+}
+pub fn cmd_verify_libraries() -> Result<()> {
+    let metadata = load_workspace()?;
+    let mut config_path = metadata.workspace_root;
+    config_path.push(".cargo/xtask.toml");
+    let config = read_xtask_toml(&config_path)?;
+
+    let mut command = Command::new("cargo")
```

Comment:
This should use `$CARGO` if available (which it will be almost all of the time, since cargo is invoked via the rustup wrapper)

Reply:
Fixed in [fa5b905](https://github.com/oxidecomputer/omicron/pull/5259/commits/fa5b905ef38a8a5f50130329967961514346ec67)

Reply:
I left it as an error if the env::var("CARGO") fails, but happy to fall back on just "cargo" if you like that better.

Reply:
Should definitely fall back to `cargo`, yeah. This pattern is pretty standard, and is expressed in bash as e.g. `${CARGO:-cargo}`.